src/fu-security-attr.c

```diff
@@ -301,17 +305,281 @@ fu_security_attrs_to_json(FuSecurityAttrs *attrs, JsonBuilder *builder)
+		fu_security_attr_dup_json(previous_json_obj, result_builder);
+		json_builder_end_object(result_builder);
+		json_builder_set_member_name(result_builder, "current");
+	} else {
+		json_builder_set_member_name(result_builder,
+					     fwupd_security_attr_get_appstream_id(current_attr));
+		json_builder_begin_object(result_builder);
+		json_builder_set_member_name(result_builder, "new");
+	}
+	json_builder_begin_object(result_builder);
+	json_builder_set_member_name(result_builder, FWUPD_RESULT_KEY_HSI_LEVEL);
+	json_builder_add_int_value(result_builder, fwupd_security_attr_get_level(current_attr));
+	json_builder_set_member_name(result_builder, FWUPD_RESULT_KEY_HSI_RESULT);
+	json_builder_add_string_value(
+	    result_builder,
+	    fwupd_security_attr_result_to_string(fwupd_security_attr_get_result(current_attr)));
+	json_builder_set_member_name(result_builder, FWUPD_RESULT_KEY_NAME);
+	json_builder_add_string_value(result_builder, fwupd_security_attr_get_name(current_attr));
+	flag_array = fu_security_attr_flag_to_string_array(current_attr);
+	if (flag_array != NULL) {
+		json_builder_set_member_name(result_builder, FWUPD_RESULT_KEY_FLAGS);
+		json_builder_begin_array(result_builder);
+		for (guint i = 0; i < flag_array->len; i++) {
+			json_builder_add_string_value(
+			    result_builder,
+			    // g_strdup(g_ptr_array_index(flag_array, i)));
+			    g_ptr_array_index(flag_array, i));
+		}
+		json_builder_end_array(result_builder);
+	}
+	json_builder_end_object(result_builder);
+	json_builder_end_object(result_builder);
+
+	return FALSE;
+}
+
+static void
+fu_security_attr_append_remove_to_result(JsonObject *previous_json_obj, JsonBuilder *result_builder)
+{
+	json_builder_set_member_name(
+	    result_builder,
+	    json_object_get_string_member(previous_json_obj, FWUPD_RESULT_KEY_APPSTREAM_ID));
+	json_builder_begin_object(result_builder);
+	json_builder_set_member_name(result_builder, "removed");
+	json_builder_begin_object(result_builder);
+	fu_security_attr_dup_json(previous_json_obj, result_builder);
+	json_builder_end_object(result_builder);
+	json_builder_end_object(result_builder);
+}
+
+gchar *
+fu_security_attrs_hsi_change(FuSecurityAttrs *attrs, const gchar *last_hsi_detail)
+{
+	g_autofree gchar *data = NULL;
+	g_autofree gchar *json_null = NULL;
+	g_autoptr(JsonParser) parser = json_parser_new();
+	g_autoptr(JsonGenerator) json_generator = NULL;
+	g_autoptr(JsonBuilder) result_builder = json_builder_new();
+	g_autoptr(JsonNode) result_json_root = NULL;
+	g_autoptr(GPtrArray) items = NULL;
+	g_autoptr(GHashTable) found_list = g_hash_table_new(g_str_hash, g_str_equal);
+	g_autoptr(GList) member_list = NULL;
+	g_autoptr(GList) removed_keys = NULL;
+	JsonNode *json_root = NULL;
+	JsonObject *json_obj = NULL;
+	JsonObject *previous_security_attrs = NULL;
+
+	if (last_hsi_detail == NULL) {
+		json_null = g_strdup("{\"SecurityAttributes\":{}}");
+		json_parser_load_from_data(parser, json_null, -1, NULL);
```

Comment:
this feels a bit like a hack; is there a better way to handle this initial case?

Reply:
I'll enhance this part. Thanks.